packages/x-date-pickers/src/internals/hooks/useField/useFieldV6TextField.ts

```diff
@@ -0,0 +1,447 @@
+import * as React from 'react';
+import { useTheme } from '@mui/material/styles';
+import useEventCallback from '@mui/utils/useEventCallback';
+import useForkRef from '@mui/utils/useForkRef';
+import { UseFieldTextFieldInteractions, UseFieldTextField } from './useField.types';
+import { FieldSection } from '../../../models';
+import { getActiveElement } from '../../utils/utils';
+import { getSectionVisibleValue, isAndroid } from './useField.utils';
+
+type FieldSectionWithPositions<TSection> = TSection & {
+  /**
+   * Start index of the section in the format
+   */
+  start: number;
+  /**
+   * End index of the section in the format
+   */
+  end: number;
+  /**
+   * Start index of the section value in the input.
+   * Takes into account invisible unicode characters such as \u2069 but does not include them
+   */
+  startInInput: number;
+  /**
+   * End index of the section value in the input.
+   * Takes into account invisible unicode characters such as \u2069 but does not include them
+   */
+  endInInput: number;
+};
+
+const cleanString = (dirtyString: string) => dirtyString.replace(/[\u2066\u2067\u2068\u2069]/g, '');
+
+export const addPositionPropertiesToSections = <TSection extends FieldSection>(
+  sections: TSection[],
+  localizedDigits: string[],
+  isRTL: boolean,
+): FieldSectionWithPositions<TSection>[] => {
+  let position = 0;
+  let positionInInput = isRTL ? 1 : 0;
+  const newSections: FieldSectionWithPositions<TSection>[] = [];
+
+  for (let i = 0; i < sections.length; i += 1) {
+    const section = sections[i];
+    const renderedValue = getSectionVisibleValue(
+      section,
+      isRTL ? 'input-rtl' : 'input-ltr',
+      localizedDigits,
+    );
+    const sectionStr = `${section.startSeparator}${renderedValue}${section.endSeparator}`;
+
+    const sectionLength = cleanString(sectionStr).length;
+    const sectionLengthInInput = sectionStr.length;
+
+    // The ...InInput values consider the unicode characters but do include them in their indexes
+    const cleanedValue = cleanString(renderedValue);
+    const startInInput =
+      positionInInput +
+      (cleanedValue === '' ? 0 : renderedValue.indexOf(cleanedValue[0])) +
+      section.startSeparator.length;
+    const endInInput = startInInput + cleanedValue.length;
+
+    newSections.push({
+      ...section,
+      start: position,
+      end: position + sectionLength,
+      startInInput,
+      endInInput,
+    });
+    position += sectionLength;
+    // Move position to the end of string associated to the current section
+    positionInInput += sectionLengthInInput;
+  }
+
+  return newSections;
+};
+
+export const useFieldV6TextField: UseFieldTextField<false> = (params) => {
+  const theme = useTheme();
+  const isRTL = theme.direction === 'rtl';
+  const focusTimeoutRef = React.useRef<NodeJS.Timeout | undefined>(undefined);
```

Comment:
This typing is already here before this PR (in `useField`).
if we set `React.useRef<number | undefined>(undefined);`, then the typing complains that `setTimeout` returns `NodeJS.Timeout`

To have `setTimeout` return a `number` you need to define the function outside of `setTimeout` (like we have in the grid `useTimeout`), not sure why, probably because the type overload works that way

But TLDR, I think we can keep it as is

Reply:
Fair points. thanks for clarifying. 🙏 🙇